samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java

```diff
@@ -106,7 +110,10 @@ public void run() {
           callback.failure(new SamzaException(msg));
         }
       };
-      ScheduledFuture scheduledFuture = timer.schedule(timerTask, timeout, TimeUnit.MILLISECONDS);
+
+      final ScheduledFuture scheduledFuture = isDraining && (envelope.isDrain() || envelope.isWatermark())
```

Comment:
why do we need both `isDraining` and `envelope.isDrain()`? What happens if the runloop has not propagated its intent but the envelope is draining? Will it be treated as regular message processing?

Reply:
Might be easier to drive this in one place as opposed to split across?
e.g., Would adding a new method w/ `createCallback(..., timeout)` fit better here? That way you can keep all the logic of drain and watermark and where we are abstracted from this piece of code (ideally thats how it should be) and then have the caller determine the timeout based on its need.